File: python/619.py
"""Project Euler Problem 619: Perfect subsets.

Find the number of non-empty subsets of {A, A+1, ... B} where the product of
the elements is a perfect square.

For each prime p, the subset must contain an even number of factors of p, which
gives an equation in mod 2 for every prime that divides at least one of the
values. If there are P such primes, then these equations can be expressed as
M ["A" "A+1" ... "B"]ᵀ = [0 0 ... 0]ᵀ, where M is a P x (B-A+1) matrix. For
large enough values of B-A, all rows of M are independent. This means that M
has rank P, and there are 2^{B-A+1 - P} solutions. Subtracting one for the
empty subset gives the answer.
"""

from __future__ import annotations

from sympy import primerange


def pow_mod(base: int, exp: int, mod: int) -> int:
    """Modular exponentiation."""
    result = 1
    base %= mod
    while exp > 0:
        if exp & 1:
            result = (result * base) % mod
        base = (base * base) % mod
        exp >>= 1
    return result


def solve() -> int:
    """Solve Problem 619."""
    A = 1_000_000
    B = 1_234_567
    M = 10**9 + 7

    rank = 0
    primes = list(primerange(2, B + 1))
    for p in primes:
        if (A - 1) // p != B // p:
            rank += 1

    ans = pow_mod(2, B - A + 1 - rank, M) - 1
    return ans % M


def main() -> int:
    """Main entry point."""
    result = solve()
    print(result)
    return result


if __name__ == "__main__":
    main()

File: c/619.c
/*
 * Project Euler Problem 619: Perfect subsets
 *
 * Count non-empty subsets of {A, A+1, ..., B} whose product is a perfect square.
 * The answer is 2^(B - A + 1 - rank) - 1 mod M, where rank is the number of
 * primes p <= B such that at least one of A..B is divisible by p.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

typedef long long ll;
typedef __int128 i128;

#define A_VAL 1000000
#define B_VAL 1234567
#define M_VAL 1000000007LL

ll pow_mod(ll base, ll exp, ll mod) {
    ll result = 1;
    base %= mod;
    if (base < 0) base += mod;
    while (exp > 0) {
        if (exp & 1) result = (i128)result * base % mod;
        base = (i128)base * base % mod;
        exp >>= 1;
    }
    return result;
}

int main(void) {
    /* Sieve primes up to B_VAL */
    char *is_prime = (char*)malloc(B_VAL + 1);
    memset(is_prime, 1, B_VAL + 1);
    is_prime[0] = is_prime[1] = 0;
    for (int i = 2; (ll)i * i <= B_VAL; i++) {
        if (is_prime[i]) {
            for (int j = i * i; j <= B_VAL; j += i)
                is_prime[j] = 0;
        }
    }

    int rank = 0;
    for (int p = 2; p <= B_VAL; p++) {
        if (is_prime[p]) {
            if ((A_VAL - 1) / p != B_VAL / p) {
                rank++;
            }
        }
    }

    ll ans = (pow_mod(2, B_VAL - A_VAL + 1 - rank, M_VAL) - 1 + M_VAL) % M_VAL;

    printf("%lld\n", ans);

    free(is_prime);
    return 0;
}
